## 🏗️ MedCLIP Architecture Diagram

```
┌─────────────────────────────────────────────────────────────────────────────┐
│                          MedCLIP Architecture [1]                           │
│                     (Frozen Encoder Training Setup)                         │
└─────────────────────────────────────────────────────────────────────────────┘

                    IMAGE INPUT                    TEXT INPUT
                 (224×224 RGB)                 (Clinical Text)
                       │                              │
                       ▼                              ▼
        ┌──────────────────────────┐  ┌──────────────────────────┐
        │   Vision Encoder (❄️)    │  │   Text Encoder (❄️)      │
        │  Swin Transformer [3]    │  │   BERT Variants [4-7]    │
        │  microsoft/swin-tiny     │  │  • BioClinicalBERT [4]   │
        │                          │  │  • PubMedBERT [5]        │
        │  27.5M params (frozen)   │  │  • BioBERT [6]           │
        │                          │  │  • BERT-base [7]         │
        │                          │  │  110M params (frozen)    │
        └───────────┬──────────────┘  └──────────────┬───────────┘
                    │                                │
                    │ [CLS] features                 │ [CLS] features
                    │ (768-dim)                      │ (768-dim)
                    ▼                                ▼
        ┌──────────────────────────┐  ┌──────────────────────────┐
        │ Vision Projection (🔥)   │  │  Text Projection (🔥)    │
        │   Linear: 768 → 512      │  │   Linear: 768 → 512      │
        │   + LayerNorm            │  │   + LayerNorm            │
        │                          │  │                          │
        │   393K params (train)    │  │   393K params (train)    │
        └───────────┬──────────────┘  └──────────────┬───────────┘
                    │                                │
                    │ v_i ∈ ℝ^512                    │ t_i ∈ ℝ^512
                    │ (L2-normalized)                │ (L2-normalized)
                    │                                │
                    └────────────┬───────────────────┘
                                 │
                                 ▼
                    ┌─────────────────────────┐
                    │   Similarity Matrix     │
                    │   S = v_i · t_j^T × τ   │
                    │   (τ = logit_scale)     │
                    │                         │
                    │   [B × B] matrix        │
                    └────────────┬────────────┘
                                 │
                                 ▼
                    ┌─────────────────────────┐
                    │ Contrastive Loss [10]   │
                    │   (InfoNCE)             │
                    │                         │
                    │ L_i2t = -log(exp(S_ii)/ │
                    │         Σ_j exp(S_ij))  │
                    │                         │
                    │ L_t2i = -log(exp(S_ii)/ │
                    │         Σ_j exp(S_ji))  │
                    │                         │
                    │ L = (L_i2t + L_t2i)/2   │
                    └─────────────────────────┘

Legend:
  ❄️  = FROZEN (weights not updated, 99.71% of params)
  🔥  = TRAINABLE (weights updated, 0.29% of params)
  B   = Batch size (512)
  τ   = Learnable temperature parameter

Training Strategy:
  • Only projection layers (786K total params) are trained
  • Vision and text encoders remain frozen
  • Batch size: 512 (large batches improve contrastive learning)
  • Learning rate: 6e-5 (scaled for large batch)
  • Optimizer: AdamW with warmup + cosine annealing
```

### Key Architectural Details:

1. **Frozen Encoders (Linear Probing)**:
   - Vision encoder: Swin Transformer pretrained on ImageNet
   - Text encoder: BERT variant pretrained on medical/general text
   - Weights are frozen to preserve pretrained knowledge [9]

2. **Trainable Projection Heads**:
   - Map 768-dim features → 512-dim shared embedding space
   - Only 393K params per encoder (0.29% of total)
   - Learn cross-modal alignment via contrastive loss

3. **Contrastive Learning**:
   - InfoNCE loss [10] pulls positive pairs together
   - Pushes negative pairs apart in embedding space
   - Symmetric loss (image→text + text→image)

4. **Multi-Encoder Comparison**:
   - Train 4 separate models with different text encoders
   - Same vision encoder across all experiments (fair comparison)
   - Evaluate which text encoder best aligns with retinal images

---

# Section 0: Environment Setup

## 0.1 Mount Google Drive

In [2]:
# Mount Google Drive for persistent storage
from google.colab import drive
import os

drive.mount('/content/drive')

# Verify mount
if os.path.exists('/content/drive/MyDrive'):
    print("✅ Google Drive mounted successfully")
else:
    print("❌ Google Drive mount failed - please check permissions")

Mounted at /content/drive
✅ Google Drive mounted successfully


## 0.2 Verify GPU

In [3]:
import torch

print("="*80)
print("🔍 GPU VERIFICATION")
print("="*80)

if torch.cuda.is_available():
    gpu_name = torch.cuda.get_device_name(0)
    gpu_mem_gb = torch.cuda.get_device_properties(0).total_memory / 1024**3

    print(f"\n✅ GPU Available!")
    print(f"   Device: {gpu_name}")
    print(f"   Memory: {gpu_mem_gb:.1f} GB")

    # Recommendations based on GPU
    if gpu_mem_gb >= 70:
        print(f"\n🎉 Excellent! A100 80GB detected")
        print(f"   → Can train 3-4 encoders in parallel")
        print(f"   → Estimated training time: ~3-4 hours")
    elif gpu_mem_gb >= 35:
        print(f"\n🎉 Great! {gpu_name}")
        print(f"   → Can train 2 encoders in parallel")
        print(f"   → Estimated training time: ~5 hours")
    elif gpu_mem_gb >= 15:
        print(f"\n⚠️ Limited VRAM")
        print(f"   → Will train sequentially (1 at a time)")
        print(f"   → Estimated training time: ~10 hours")
    else:
        print(f"\n❌ Insufficient VRAM")
        print(f"   → Need at least 16GB for training")
        print(f"   → Please upgrade to T4, V100, or A100")
else:
    print(f"\n❌ NO GPU DETECTED!")
    print(f"\n⚠️ CRITICAL: This notebook requires a GPU")
    print(f"\nHow to fix:")
    print(f"   1. Click 'Runtime' → 'Change runtime type'")
    print(f"   2. Set 'Hardware accelerator' to 'GPU'")
    print(f"   3. Select 'GPU type': A100 (recommended)")
    print(f"   4. Click 'Save' and re-run this cell")

print("\n" + "="*80)

🔍 GPU VERIFICATION

✅ GPU Available!
   Device: NVIDIA L4
   Memory: 22.2 GB

⚠️ Limited VRAM
   → Will train sequentially (1 at a time)
   → Estimated training time: ~10 hours



## 0.3 Global Configuration

In [4]:
import os

# ============================================================================
# GLOBAL CONFIGURATION
# ============================================================================

# Google Drive Paths
DRIVE_BASE = "/content/drive/MyDrive/MEDCLIP-PEACEIN"
DRIVE_DATA = f"{DRIVE_BASE}/data"
DRIVE_CHECKPOINTS = f"{DRIVE_BASE}/checkpoints"
DRIVE_RESULTS = f"{DRIVE_BASE}/results"


# Text Encoders to Compare (4 total: MedCLIP default + medical variants + general baseline)
TEXT_ENCODERS = [
    {
        "name": "BioClinicalBERT",
        "hf_id": "emilyalsentzer/Bio_ClinicalBERT",
        "description": "Clinical notes (MIMIC-III) [MedCLIP Default]",
        "expected_rank": 1
    },
    {
        "name": "PubMedBERT",
        "hf_id": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
        "description": "Medical abstracts + full-text",
        "expected_rank": 2
    },
    {
        "name": "BioBERT",
        "hf_id": "dmis-lab/biobert-base-cased-v1.1",
        "description": "PubMed + PMC articles",
        "expected_rank": 3
    },
    {
        "name": "BERT-base",
        "hf_id": "bert-base-uncased",
        "description": "General English baseline (no medical)",
        "expected_rank": 4
    }
]

# Training Hyperparameters (optimized for A100 80GB)
LEARNING_RATE = 6e-5  # Scaled up for larger batch size
BATCH_SIZE = 512  # Increased for A100 80GB (was 128)
NUM_EPOCHS = 20
WARMUP_STEPS = 50
MAX_GRAD_NORM = 1.0
WEIGHT_DECAY = 0.01
IMAGE_SIZE = 224
CONTEXT_LENGTH = 77  # MedCLIP default

# Dataset Configuration
DATASET_NAME = "Peacein/color-fundus-eye"
DISEASE_LABELS = [
    "Central Serous Chorioretinopathy",
    "Diabetic Retinopathy",
    "Disc Edema",
    "Glaucoma",
    "Healthy",
    "Macular Scar",
    "Myopia",
    "Pterygium",
    "Retinal Detachment",
    "Retinitis Pigmentosa"
]

# DSPy Configuration (for prompt generation if needed)
PRIMARY_MODEL = "openrouter/google/gemini-2.5-flash-lite-preview-09-2025"
DELAY_BETWEEN_CALLS = 1.0
CHECKPOINT_INTERVAL = 10

# Create directories
for path in [DRIVE_BASE, DRIVE_DATA, DRIVE_CHECKPOINTS, DRIVE_RESULTS]:
    os.makedirs(path, exist_ok=True)

# Print configuration summary
print("="*80)
print("📊 CONFIGURATION SUMMARY")
print("="*80)

print(f"\n💾 Paths:")
print(f"   Google Drive: {DRIVE_BASE}")

print(f"\n🔤 Text Encoders ({len(TEXT_ENCODERS)}):")
for enc in TEXT_ENCODERS:
    print(f"   {enc['expected_rank']}. {enc['name']:20s} - {enc['description']}")

print(f"\n⚙️ Training Hyperparameters (A100 80GB Optimized):")
print(f"   Learning rate: {LEARNING_RATE} (scaled for batch size)")
print(f"   Batch size: {BATCH_SIZE} ⚡ (4x larger for A100)")
print(f"   Epochs: {NUM_EPOCHS}")
print(f"   Image size: {IMAGE_SIZE}x{IMAGE_SIZE}")
print(f"   Max grad norm: {MAX_GRAD_NORM}")

print(f"\n🏷️ Disease Classes ({len(DISEASE_LABELS)}):")
for i, label in enumerate(DISEASE_LABELS, 1):
    print(f"   {i}. {label}")

print("\n" + "="*80)

📊 CONFIGURATION SUMMARY

💾 Paths:
   Google Drive: /content/drive/MyDrive/MEDCLIP-PEACEIN

🔤 Text Encoders (4):
   1. BioClinicalBERT      - Clinical notes (MIMIC-III) [MedCLIP Default]
   2. PubMedBERT           - Medical abstracts + full-text
   3. BioBERT              - PubMed + PMC articles
   4. BERT-base            - General English baseline (no medical)

⚙️ Training Hyperparameters (A100 80GB Optimized):
   Learning rate: 6e-05 (scaled for batch size)
   Batch size: 512 ⚡ (4x larger for A100)
   Epochs: 20
   Image size: 224x224
   Max grad norm: 1.0

🏷️ Disease Classes (10):
   1. Central Serous Chorioretinopathy
   2. Diabetic Retinopathy
   3. Disc Edema
   4. Glaucoma
   5. Healthy
   6. Macular Scar
   7. Myopia
   8. Pterygium
   9. Retinal Detachment
   10. Retinitis Pigmentosa



---

# Section 1: Check for Existing Data Files ✨

**Following ODIR Pattern**: Check if TSV/JSONL files already exist before loading the full dataset.

This is the **CRITICAL OPTIMIZATION** that saves ~9-10 hours if you've already generated prompts!

In [5]:
import os
import pandas as pd
import json

print("="*80)
print("🔍 CHECKING FOR EXISTING DATA FILES")
print("="*80)

# Define file paths (UPDATED: Look in Google Drive instead of repo)
train_imgs_tsv = f"{DRIVE_DATA}/train_imgs.tsv"
train_texts_jsonl = f"{DRIVE_DATA}/train_texts.jsonl"
test_imgs_tsv = f"{DRIVE_DATA}/test_imgs.tsv"
test_texts_jsonl = f"{DRIVE_DATA}/test_texts.jsonl"

# Check if all files exist
files_status = {
    "Train Images TSV": os.path.exists(train_imgs_tsv),
    "Train Texts JSONL": os.path.exists(train_texts_jsonl),
    "Test Images TSV": os.path.exists(test_imgs_tsv),
    "Test Texts JSONL": os.path.exists(test_texts_jsonl)
}

print(f"\n📋 File Status (Google Drive):")
print(f"   Location: {DRIVE_DATA}")
for file_name, exists in files_status.items():
    status = "✅ FOUND" if exists else "❌ MISSING"
    print(f"   {file_name:20s}: {status}")

USE_EXISTING_DATA = all(files_status.values())

print("\n" + "="*80)
if USE_EXISTING_DATA:
    print("🎉 ALL FILES FOUND - USING EXISTING DATA")
    print("="*80)
    print("\n⚡ FAST PATH ACTIVATED:")
    print("   ⏭️ Skipping HuggingFace dataset download")
    print("   ⏭️ Skipping DSPy prompt generation (~9-10 hours saved!)")
    print("\n👉 You can skip directly to Section 4 (MedCLIP Setup)")

    # Load and verify files
    print(f"\n📊 Verifying data integrity...")

    try:
        # Load TSV files
        train_imgs_df = pd.read_csv(train_imgs_tsv, sep="\t", header=None,
                                     names=["patient_id", "base64_img"])
        test_imgs_df = pd.read_csv(test_imgs_tsv, sep="\t", header=None,
                                    names=["patient_id", "base64_img"])

        # Load JSONL files
        with open(train_texts_jsonl) as f:
            train_texts = [json.loads(line) for line in f]
        with open(test_texts_jsonl) as f:
            test_texts = [json.loads(line) for line in f]

        print(f"\n✅ Data Loaded Successfully:")
        print(f"   Train images: {len(train_imgs_df):,}")
        print(f"   Train prompts: {len(train_texts):,}")
        print(f"   Test images: {len(test_imgs_df):,}")
        print(f"   Test prompts: {len(test_texts):,}")

        # Preview
        print(f"\n📋 Sample Data:")
        print(f"   First patient ID: {train_imgs_df.iloc[0]['patient_id']}")
        print(f"   First prompt (truncated): {train_texts[0]['text'][:100]}...")

    except Exception as e:
        print(f"\n❌ Error loading files: {e}")
        print("   Files may be corrupted. Will regenerate in Sections 2-3.")
        USE_EXISTING_DATA = False

else:
    print("⚠️ FILES MISSING - WILL GENERATE DATA")
    print("="*80)
    print("\n🐌 SLOW PATH:")
    print("   📥 Will download Peacein dataset from HuggingFace")
    print("   ✍️ Will generate prompts using DSPy (~9-10 hours)")
    print("   💾 Will save TSV/JSONL files to Google Drive for future runs")
    print(f"\n📍 Target location: {DRIVE_DATA}")
    print("\n👉 Continue to Section 2 to start data generation")

print("\n" + "="*80)

🔍 CHECKING FOR EXISTING DATA FILES

📋 File Status (Google Drive):
   Location: /content/drive/MyDrive/MEDCLIP-PEACEIN/data
   Train Images TSV    : ✅ FOUND
   Train Texts JSONL   : ✅ FOUND
   Test Images TSV     : ✅ FOUND
   Test Texts JSONL    : ✅ FOUND

🎉 ALL FILES FOUND - USING EXISTING DATA

⚡ FAST PATH ACTIVATED:
   ⏭️ Skipping HuggingFace dataset download
   ⏭️ Skipping DSPy prompt generation (~9-10 hours saved!)

👉 You can skip directly to Section 4 (MedCLIP Setup)

📊 Verifying data integrity...

✅ Data Loaded Successfully:
   Train images: 12,854
   Train prompts: 12,854
   Test images: 3,253
   Test prompts: 3,253

📋 Sample Data:
   First patient ID: _9j_4AAQSkZJRgABAQAAAQABAAD_2wBDAAIBAQEBAQIBAQECAgICAgQDAgICAgUEBAMEBgUGBgYFBgYGBwkIBgcJBwYGCAsICQoKCgoKBggLDAsKDAkKCgr_2wBDAQICAgICAgUDAwUKBwYHCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgr_wAARCADgAOADASIAAhEBAxEB_8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL_8QAtRAAAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJx

---

# Section 2: Dataset Loading (Only if files missing)

**⚠️ SKIP THIS SECTION IF YOU HAVE EXISTING DATA!**

This section downloads the Peacein dataset from HuggingFace.

## 2.1 Load Peacein Dataset from HuggingFace

In [5]:
# Only run if USE_EXISTING_DATA is False
if not USE_EXISTING_DATA:
    from datasets import load_dataset
    import pandas as pd

    print("="*80)
    print("📥 LOADING PEACEIN DATASET")
    print("="*80)

    # Load dataset
    print(f"\nLoading {DATASET_NAME}...")
    dataset = load_dataset(DATASET_NAME)

    print(f"\n✅ Dataset loaded!")
    print(f"   Train: {len(dataset['train']):,} samples")
    print(f"   Test: {len(dataset['test']):,} samples")

    # Convert to DataFrames for easier manipulation
    train_df = dataset['train'].to_pandas()
    test_df = dataset['test'].to_pandas()

    print(f"\n📊 Dataset info:")
    print(f"   Columns: {list(train_df.columns)}")
    print(f"   Image shape: {train_df['image'][0].size}")
    print(f"   Labels: {train_df['label'].unique()}")

    # Map label indices to disease names
    label_to_disease = dict(enumerate(DISEASE_LABELS))
    train_df['disease'] = train_df['label'].map(label_to_disease)
    test_df['disease'] = test_df['label'].map(label_to_disease)

    print("\n" + "="*80)
else:
    print("⏭️ Skipping Section 2: Using existing data")

⏭️ Skipping Section 2: Using existing data


---

# Section 3: DSPy Prompt Generation (Only if files missing)

**⚠️ SKIP THIS SECTION IF YOU HAVE EXISTING PROMPTS!**

Generate clinical diagnostic impressions using DSPy.

## 3.1 Setup DSPy

In [6]:
if not USE_EXISTING_DATA:
    # Install DSPy
    !pip install -q dspy-ai

    import dspy
    import os

    # Configure DSPy with OpenRouter
    lm = dspy.LM(PRIMARY_MODEL, api_key=os.environ.get('OPENROUTER_API_KEY'))
    dspy.configure(lm=lm)

    print("✅ DSPy configured")
    print(f"   Model: {PRIMARY_MODEL}")
else:
    print("⏭️ Skipping Section 3: Using existing prompts")

⏭️ Skipping Section 3: Using existing prompts


## 3.2 Define Prompt Generation Signature

In [7]:
if not USE_EXISTING_DATA:
    class GenerateDiagnosticImpression(dspy.Signature):
        """Generate a clinical diagnostic impression for a fundus photograph."""

        disease_name = dspy.InputField(desc="The diagnosed retinal disease")
        diagnostic_impression = dspy.OutputField(
            desc="A clinical diagnostic impression (2-3 sentences) describing observable features"
        )

    # Create predictor
    predictor = dspy.Predict(GenerateDiagnosticImpression)

    print("✅ DSPy signature defined")
else:
    print("⏭️ Skipping DSPy setup")

⏭️ Skipping DSPy setup


## 3.3 Generate Prompts for All Images

In [8]:
if not USE_EXISTING_DATA:
    from tqdm.notebook import tqdm
    import time

    print("="*80)
    print("✍️ GENERATING PROMPTS")
    print("="*80)

    def generate_prompt(disease_name):
        """Generate a single diagnostic impression."""
        try:
            result = predictor(disease_name=disease_name)
            return result.diagnostic_impression
        except Exception as e:
            print(f"   Error: {e}")
            return f"Fundus photograph showing {disease_name}."

    # Generate for train set
    print(f"\n📝 Generating prompts for train set ({len(train_df)} samples)...")
    train_prompts = []
    for idx, row in tqdm(train_df.iterrows(), total=len(train_df)):
        prompt = generate_prompt(row['disease'])
        train_prompts.append(prompt)
        time.sleep(DELAY_BETWEEN_CALLS)

        # Checkpoint
        if (idx + 1) % CHECKPOINT_INTERVAL == 0:
            print(f"   Checkpoint: {idx + 1}/{len(train_df)}")

    train_df['prompt'] = train_prompts

    # Generate for test set
    print(f"\n📝 Generating prompts for test set ({len(test_df)} samples)...")
    test_prompts = []
    for idx, row in tqdm(test_df.iterrows(), total=len(test_df)):
        prompt = generate_prompt(row['disease'])
        test_prompts.append(prompt)
        time.sleep(DELAY_BETWEEN_CALLS)

    test_df['prompt'] = test_prompts

    print(f"\n✅ Prompts generated!")
    print(f"   Train: {len(train_prompts)}")
    print(f"   Test: {len(test_prompts)}")
    print("\n" + "="*80)
else:
    print("⏭️ Skipping prompt generation")

⏭️ Skipping prompt generation


## 3.4 Save as TSV/JSONL

In [9]:
if not USE_EXISTING_DATA:
    import base64
    from io import BytesIO

    print("="*80)
    print("💾 SAVING TSV/JSONL FILES")
    print("="*80)

    def image_to_base64(pil_image):
        """Convert PIL Image to base64 string."""
        buffered = BytesIO()
        pil_image.save(buffered, format="JPEG")
        img_str = base64.b64encode(buffered.getvalue()).decode()
        return img_str.replace('/', '_')  # Replace / with _ for TSV safety

    # Create train TSV
    print(f"\n📄 Creating train_imgs.tsv...")
    train_imgs_data = []
    for idx, row in tqdm(train_df.iterrows(), total=len(train_df)):
        patient_id = f"patient_{idx:06d}"
        base64_img = image_to_base64(row['image'])
        train_imgs_data.append({'patient_id': patient_id, 'base64_img': base64_img})

    train_imgs_df = pd.DataFrame(train_imgs_data)
    train_imgs_df.to_csv(f"{DRIVE_DATA}/train_imgs.tsv", sep='\t', header=False, index=False)

    # Create train JSONL
    print(f"\n📄 Creating train_texts.jsonl...")
    with open(f"{DRIVE_DATA}/train_texts.jsonl", 'w') as f:
        for idx, row in train_df.iterrows():
            patient_id = f"patient_{idx:06d}"
            entry = {
                'text_id': idx,
                'text': row['prompt'],
                'image_ids': [patient_id],
                'label': int(row['label']),
                'disease': row['disease']
            }
            f.write(json.dumps(entry) + '\n')

    # Create test TSV
    print(f"\n📄 Creating test_imgs.tsv...")
    test_imgs_data = []
    for idx, row in tqdm(test_df.iterrows(), total=len(test_df)):
        patient_id = f"patient_{idx:06d}"
        base64_img = image_to_base64(row['image'])
        test_imgs_data.append({'patient_id': patient_id, 'base64_img': base64_img})

    test_imgs_df = pd.DataFrame(test_imgs_data)
    test_imgs_df.to_csv(f"{DRIVE_DATA}/test_imgs.tsv", sep='\t', header=False, index=False)

    # Create test JSONL
    print(f"\n📄 Creating test_texts.jsonl...")
    with open(f"{DRIVE_DATA}/test_texts.jsonl", 'w') as f:
        for idx, row in test_df.iterrows():
            patient_id = f"patient_{idx:06d}"
            entry = {
                'text_id': idx,
                'text': row['prompt'],
                'image_ids': [patient_id],
                'label': int(row['label']),
                'disease': row['disease']
            }
            f.write(json.dumps(entry) + '\n')

    print(f"\n✅ All files saved to {DRIVE_DATA}")
    print("\n" + "="*80)

    # Update variables for next sections
    train_imgs_tsv = f"{DRIVE_DATA}/train_imgs.tsv"
    train_texts_jsonl = f"{DRIVE_DATA}/train_texts.jsonl"
    test_imgs_tsv = f"{DRIVE_DATA}/test_imgs.tsv"
    test_texts_jsonl = f"{DRIVE_DATA}/test_texts.jsonl"
    USE_EXISTING_DATA = True
else:
    print("⏭️ Skipping file creation")

⏭️ Skipping file creation


---

# Section 4: Clone & Modify MedCLIP 🔧

**Critical**: Clone MedCLIP and make text encoder swappable.

## 4.1 Clone MedCLIP

In [10]:
import os
import subprocess

MEDCLIP_REPO = "https://github.com/RyanWangZf/MedCLIP.git"
MEDCLIP_DIR = "/content/MedCLIP"

print("="*80)
print("📥 CLONING MEDCLIP")
print("="*80)

if os.path.exists(MEDCLIP_DIR):
    print(f"\n✅ Already exists at {MEDCLIP_DIR}")
else:
    print(f"\n📥 Cloning...")
    subprocess.run(["git", "clone", MEDCLIP_REPO, MEDCLIP_DIR], check=True)
    print("✅ Cloned")

print(f"\n📂 Contents: {', '.join(os.listdir(MEDCLIP_DIR)[:8])}")
print("\n" + "="*80)

📥 CLONING MEDCLIP

📥 Cloning...
✅ Cloned

📂 Contents: local_data, medclip, example_data, examples, pypi_build_commands.txt, .git, requirements.txt, .gitignore



## 4.2 Modify constants.py for Swappable Encoders

In [11]:
# Modify MedCLIP constants.py to allow dynamic BERT_TYPE
import os

constants_path = f"{MEDCLIP_DIR}/medclip/constants.py"

# Append helper function to change BERT_TYPE
with open(constants_path, 'a') as f:
    f.write("""

# Added for multi-encoder comparison
def set_bert_type(model_name):
    global BERT_TYPE
    BERT_TYPE = model_name
    print(f"✅ BERT_TYPE updated to: {BERT_TYPE}")
""")

print("🔧 Modified constants.py to support encoder swapping")
print(f"   File: {constants_path}")

🔧 Modified constants.py to support encoder swapping
   File: /content/MedCLIP/medclip/constants.py


## 4.3 Install MedCLIP

In [12]:
!cd {MEDCLIP_DIR} && pip install -q -e .

import torch
import transformers
print("✅ MedCLIP installed")
print(f"   PyTorch: {torch.__version__}")
print(f"   Transformers: {transformers.__version__}")
print(f"   CUDA: {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"   GPU: {torch.cuda.get_device_name(0)}")

  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.5/90.5 kB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 314.9/314.9 kB 27.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 111.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.9/27.9 MB 100.6 MB/s eta 0:00:00
  error: subprocess-exited-with-error
  
  × Building wheel for tokenizers (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for tokenizers
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (tokenizers)
✅ MedCLIP installed
   PyTorch: 2.9.0+cu126
   Transformers: 4.5

---

# Section 5: Data Preprocessing

Load TSV/JSONL and create PyTorch datasets.

## 5.1 Custom Dataset Class

In [13]:
import sys
sys.path.insert(0, MEDCLIP_DIR)

import torch
from torch.utils.data import Dataset, DataLoader
import pandas as pd
import json
import base64
from io import BytesIO
from PIL import Image
import numpy as np
from medclip import MedCLIPProcessor

# CRITICAL FIX: Patch MedCLIP processor bugs
from medclip.dataset import MedCLIPFeatureExtractor

def _convert_rgb(self, image):
    if isinstance(image, Image.Image):
        if image.mode != 'RGB':
            return image.convert('RGB')
        return image
    return image

def _pad_img_fixed(self, img, min_size=224, fill_color=0):
    if isinstance(min_size, dict):
        min_size = min_size.get('height', min_size.get('shortest_edge', 224))

    if isinstance(img, np.ndarray):
        if len(img.shape) == 3:
            y, x = img.shape[:2]
        else:
            y, x = img.shape
        img = Image.fromarray(img)
    else:
        x, y = img.size

    size = max(min_size, x, y)

    if img.mode == 'RGB':
        mode = 'RGB'
        fill = (fill_color, fill_color, fill_color)
    else:
        mode = 'L'
        fill = fill_color

    new_im = Image.new(mode, (size, size), fill)
    new_im.paste(img, (int((size - x) / 2), int((size - y) / 2)))

    img_array = np.array(new_im)
    if len(img_array.shape) == 3:
        img_array = np.transpose(img_array, (2, 0, 1))

    return img_array

MedCLIPFeatureExtractor.convert_rgb = _convert_rgb
MedCLIPFeatureExtractor.pad_img = _pad_img_fixed
print("✅ Patched MedCLIPFeatureExtractor")

class PeaceinMedCLIPDataset(Dataset):
    def __init__(self, imgs_tsv_path, texts_jsonl_path, processor):
        self.imgs_df = pd.read_csv(imgs_tsv_path, sep="\t", header=None,
                                    names=["patient_id", "base64_img"])
        with open(texts_jsonl_path) as f:
            self.texts = [json.loads(line) for line in f]
        self.processor = processor

    def __len__(self):
        return min(len(self.texts), len(self.imgs_df))

    def _decode_base64_image(self, b64_string):
        b64_string = ''.join(b64_string.split())
        b64_string = b64_string.replace('_', '/').replace('-', '+')
        missing_padding = len(b64_string) % 4
        if missing_padding:
            b64_string += '=' * (4 - missing_padding)
        return base64.b64decode(b64_string)

    def __getitem__(self, idx):
        text_entry = self.texts[idx]
        img_b64 = self.imgs_df.iloc[idx]['base64_img']
        img_bytes = self._decode_base64_image(img_b64)
        image = Image.open(BytesIO(img_bytes)).convert('RGB')

        # Ensure text is clean string
        text = str(text_entry['text']).strip()

        return {'image': image, 'text': text}

# Create global processor to reuse
_global_processor = None

def medclip_collate_fn(batch):
    global _global_processor
    if _global_processor is None:
        _global_processor = MedCLIPProcessor()

    images = [item['image'] for item in batch]
    texts = [item['text'] for item in batch]

    # Process images
    image_inputs = _global_processor.feature_extractor(images=images, return_tensors="pt")

    # Tokenize using __call__ with list of strings
    from transformers import AutoTokenizer
    tokenizer = AutoTokenizer.from_pretrained(_global_processor.tokenizer.name_or_path)

    text_inputs = tokenizer(
        texts,
        padding='max_length',
        max_length=CONTEXT_LENGTH,
        truncation=True,
        return_tensors='pt'
    )

    return {
        'input_ids': text_inputs['input_ids'],
        'attention_mask': text_inputs['attention_mask'],
        'pixel_values': image_inputs['pixel_values']
    }

print("✅ Dataset and collate_fn defined")

/content/MedCLIP/medclip/dataset.py:265: SyntaxWarning: invalid escape sequence '\.'
  splitter = re.compile("[0-9]+\.+[^0-9]")


✅ Patched MedCLIPFeatureExtractor
✅ Dataset and collate_fn defined


## 5.2 Create Datasets

In [14]:
processor = MedCLIPProcessor()

train_dataset = PeaceinMedCLIPDataset(
    imgs_tsv_path=train_imgs_tsv,
    texts_jsonl_path=train_texts_jsonl,
    processor=processor
)

test_dataset = PeaceinMedCLIPDataset(
    imgs_tsv_path=test_imgs_tsv,
    texts_jsonl_path=test_texts_jsonl,
    processor=processor
)

print(f"✅ Datasets created")
print(f"   Train: {len(train_dataset):,} samples")
print(f"   Test: {len(test_dataset):,} samples")

# Test loading one sample with collate_fn
sample_batch = medclip_collate_fn([train_dataset[0]])
print(f"\n📋 Sample batch shapes:")
print(f"   input_ids: {sample_batch['input_ids'].shape}")
print(f"   attention_mask: {sample_batch['attention_mask'].shape}")
print(f"   pixel_values: {sample_batch['pixel_values'].shape}")

/usr/local/lib/python3.12/dist-packages/transformers/models/clip/feature_extraction_clip.py:30: FutureWarning: The class CLIPFeatureExtractor is deprecated and will be removed in version 5 of Transformers. Please use CLIPImageProcessor instead.
  warnings.warn(


config.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

✅ Datasets created
   Train: 12,854 samples
   Test: 3,253 samples


/usr/local/lib/python3.12/dist-packages/transformers/models/clip/feature_extraction_clip.py:30: FutureWarning: The class CLIPFeatureExtractor is deprecated and will be removed in version 5 of Transformers. Please use CLIPImageProcessor instead.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/transformers/models/clip/processing_clip.py:66: FutureWarning: `feature_extractor` is deprecated and will be removed in v5. Use `image_processor` instead.
  warnings.warn(



📋 Sample batch shapes:
   input_ids: torch.Size([1, 77])
   attention_mask: torch.Size([1, 77])
   pixel_values: torch.Size([1, 3, 224, 224])


---

# Section 6: Multi-Encoder FROZEN Training Loop 🔒

**CRITICAL**: Train projection layers while keeping encoders FROZEN.

This loop trains all 4 encoders sequentially.

In [20]:
import sys
sys.path.insert(0, MEDCLIP_DIR)

from medclip import MedCLIPModel, MedCLIPVisionModelViT
from medclip.losses import ImageTextContrastiveLoss
from medclip.trainer import Trainer
import medclip.constants as constants
import torch
from torch.utils.data import DataLoader, Subset
import time
import os
import numpy as np
from sklearn.metrics import accuracy_score

# ============================================================================
# VALIDATION EVALUATOR CLASS (FIXED: collect embeddings first!)
# ============================================================================

class MedCLIPEvaluator:
    """Simple evaluator for monitoring validation accuracy during training."""

    def __init__(self, model, val_dataloader, device='cuda'):
        self.model = model
        self.val_dataloader = val_dataloader
        self.device = device

    def evaluate(self):
        """Run validation and return metrics."""
        self.model.eval()
        all_img_embeds = []
        all_text_embeds = []

        with torch.no_grad():
            for batch in self.val_dataloader:
                input_ids = batch['input_ids'].to(self.device)
                attention_mask = batch['attention_mask'].to(self.device)
                pixel_values = batch['pixel_values'].to(self.device)

                # Get embeddings
                img_embeds = self.model.encode_image(pixel_values)
                text_embeds = self.model.encode_text(input_ids, attention_mask)

                # Collect embeddings (NOT similarities!)
                all_img_embeds.append(img_embeds.cpu())
                all_text_embeds.append(text_embeds.cpu())

        # Concatenate all embeddings
        all_img_embeds = torch.cat(all_img_embeds, dim=0)
        all_text_embeds = torch.cat(all_text_embeds, dim=0)

        # NOW compute full similarity matrix
        similarity = torch.matmul(all_img_embeds, all_text_embeds.T)

        # Accuracy: correct if diagonal is max in row
        predictions = similarity.argmax(dim=1)
        correct = (predictions == torch.arange(len(predictions))).float()
        accuracy = correct.mean().item()

        self.model.train()

        return {
            'acc': accuracy,
            'auc': accuracy  # Dummy value (trainer expects this)
        }


# ============================================================================
# TRAINING FUNCTION WITH VALIDATION
# ============================================================================

def train_single_encoder(encoder_config, train_dataset, val_split=0.1):
    """Train a single encoder with frozen base models + validation monitoring.

    Args:
        encoder_config: Dict with 'name' and 'hf_id'
        train_dataset: Full training dataset
        val_split: Fraction of data to use for validation (default: 0.1 = 10%)
    """
    encoder_name = encoder_config['name']
    encoder_hf_id = encoder_config['hf_id']

    print(f"\n{'='*80}")
    print(f"🚀 TRAINING: {encoder_name}")
    print(f"{'='*80}")

    # Set text encoder dynamically
    constants.set_bert_type(encoder_hf_id)

    # Initialize model
    print(f"\n📥 Loading model...")
    model = MedCLIPModel(vision_cls=MedCLIPVisionModelViT)
    model.cuda()

    # ============================================================================
    # 🔒 FREEZE ENCODERS (FIXED!) - Unfreeze projections for domain adaptation
    # ============================================================================
    print(f"\n🔒 Freezing encoders (keeping projections trainable)...")

    # ❄️ Freeze vision encoder (Swin Transformer) - NOT the projection!
    for param in model.vision_model.model.parameters():
        param.requires_grad = False

    # ❄️ Freeze text encoder (BERT) - NOT the projection!
    for param in model.text_model.model.parameters():
        param.requires_grad = False

    # 🔥 Explicitly unfreeze projection heads (for domain adaptation: CXR → fundus)
    for param in model.vision_model.projection_head.parameters():
        param.requires_grad = True
    for param in model.text_model.projection_head.parameters():
        param.requires_grad = True

    # Count trainable vs frozen
    trainable = sum(p.numel() for p in model.parameters() if p.requires_grad)
    total = sum(p.numel() for p in model.parameters())

    print(f"   Trainable: {trainable:,} ({trainable/total*100:.2f}%)")
    print(f"   Frozen: {total-trainable:,} ({(total-trainable)/total*100:.2f}%)")

    # Verify correct parts are trainable
    vision_proj_trainable = sum(p.numel() for p in model.vision_model.projection_head.parameters() if p.requires_grad)
    text_proj_trainable = sum(p.numel() for p in model.text_model.projection_head.parameters() if p.requires_grad)
    print(f"   Vision projection: {vision_proj_trainable:,} trainable ✅")
    print(f"   Text projection: {text_proj_trainable:,} trainable ✅")
    print(f"   Logit scale: trainable ✅")

    # ============================================================================
    # VALIDATION SPLIT (10% of training data)
    # ============================================================================
    print(f"\n📊 Creating train/validation split ({int(val_split*100)}% val)...")

    n_samples = len(train_dataset)
    indices = np.random.RandomState(42).permutation(n_samples)

    n_val = int(n_samples * val_split)
    val_indices = indices[:n_val]
    train_indices = indices[n_val:]

    train_subset = Subset(train_dataset, train_indices)
    val_subset = Subset(train_dataset, val_indices)

    print(f"   Train: {len(train_subset):,} samples")
    print(f"   Val: {len(val_subset):,} samples")

    # ============================================================================
    # DATALOADERS + EVALUATOR
    # ============================================================================
    print(f"\n⚙️ Preparing training...")

    # Calculate steps per epoch for evaluation frequency
    steps_per_epoch = len(train_subset) // BATCH_SIZE
    print(f"   Steps per epoch: {steps_per_epoch}")

    train_dataloader = DataLoader(
        train_subset,
        batch_size=BATCH_SIZE,
        shuffle=True,
        num_workers=0,
        pin_memory=True,
        collate_fn=medclip_collate_fn
    )

    val_dataloader = DataLoader(
        val_subset,
        batch_size=64,  # Smaller batch for validation
        shuffle=False,
        num_workers=0,
        pin_memory=True,
        collate_fn=medclip_collate_fn
    )

    # Create evaluator
    evaluator = MedCLIPEvaluator(model, val_dataloader)

    # Loss model
    loss_model = ImageTextContrastiveLoss(model=model)

    # Checkpoint directory
    checkpoint_dir = f"{DRIVE_CHECKPOINTS}/{encoder_name.replace('/', '-')}"
    os.makedirs(checkpoint_dir, exist_ok=True)

    # ============================================================================
    # TRAINING
    # ============================================================================
    trainer = Trainer()

    print(f"\n🎯 Starting training ({NUM_EPOCHS} epochs)...")
    print(f"   Batch size: {BATCH_SIZE}")
    print(f"   Learning rate: {LEARNING_RATE}")
    print(f"   Validation every {steps_per_epoch} steps (every epoch)")
    print(f"   Checkpoints every {steps_per_epoch} steps (every epoch)")
    start_time = time.time()

    trainer.train(
        model=model,
        train_objectives=[(train_dataloader, loss_model, 1.0)],
        eval_dataloader=val_dataloader,
        evaluator=evaluator,
        epochs=NUM_EPOCHS,
        optimizer_params={'lr': LEARNING_RATE},
        warmup_steps=WARMUP_STEPS,
        max_grad_norm=MAX_GRAD_NORM,
        weight_decay=WEIGHT_DECAY,
        output_path=checkpoint_dir,
        save_steps=steps_per_epoch,
        evaluation_steps=steps_per_epoch,
        save_best_model=True,        # Save best based on validation
        load_best_model_at_last=True # Load best at end
    )

    training_time = time.time() - start_time

    print(f"\n✅ {encoder_name} complete!")
    print(f"   Time: {training_time/3600:.2f}h")
    print(f"   Best model saved to: {checkpoint_dir}/best")

    # Use best model for final checkpoint
    final_path = f"{checkpoint_dir}/best"

    # Clear GPU
    del model, trainer, loss_model, evaluator
    torch.cuda.empty_cache()

    return {
        'encoder_name': encoder_name,
        'training_time_hours': training_time / 3600,
        'checkpoint_path': final_path
    }


# ============================================================================
# TRAIN ALL ENCODERS SEQUENTIALLY
# ============================================================================

print("="*80)
print("🚀 TRAINING ALL ENCODERS")
print("="*80)

all_training_results = {}

for encoder_config in TEXT_ENCODERS:
    result = train_single_encoder(encoder_config, train_dataset, val_split=0.1)
    all_training_results[result['encoder_name']] = result

print(f"\n\n{'='*80}")
print(f"🎉 ALL ENCODERS TRAINED!")
print(f"{'='*80}")

total_time = sum(r['training_time_hours'] for r in all_training_results.values())
print(f"\n📊 Summary:")
for name, results in all_training_results.items():
    print(f"   {name:20s}: {results['training_time_hours']:.2f}h")
print(f"\n   Total: {total_time:.2f}h")

# Store for Section 7
training_results = all_training_results

🚀 TRAINING ALL ENCODERS

🚀 TRAINING: BioClinicalBERT
✅ BERT_TYPE updated to: emilyalsentzer/Bio_ClinicalBERT

📥 Loading model...

🔒 Freezing encoders (keeping projections trainable)...
   Trainable: 786,433 (0.58%)
   Frozen: 135,829,626 (99.42%)
   Vision projection: 393,216 trainable ✅
   Text projection: 393,216 trainable ✅
   Logit scale: trainable ✅

📊 Creating train/validation split (10% val)...
   Train: 11,569 samples
   Val: 1,285 samples

⚙️ Preparing training...
   Steps per epoch: 22

🎯 Starting training (20 epochs)...
   Batch size: 512
   Learning rate: 6e-05
   Validation every 22 steps (every epoch)
   Checkpoints every 22 steps (every epoch)


Epoch:   0%|          | 0/20 [00:00<?, ?it/s]

Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 6.2271 


######### Eval 22 #########
acc: 0.0016
auc: 0.0016


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 6.1002 


######### Eval 44 #########
acc: 0.0008
auc: 0.0008


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.9986 


######### Eval 66 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.9174 


######### Eval 88 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.8307 


######### Eval 110 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7675 


######### Eval 132 #########
acc: 0.0062
auc: 0.0062


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7323 


######### Eval 154 #########
acc: 0.0062
auc: 0.0062


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6903 


######### Eval 176 #########
acc: 0.0078
auc: 0.0078


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6665 


######### Eval 198 #########
acc: 0.0062
auc: 0.0062


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6524 


######### Eval 220 #########
acc: 0.0070
auc: 0.0070


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6247 


######### Eval 242 #########
acc: 0.0062
auc: 0.0062


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6097 


######### Eval 264 #########
acc: 0.0078
auc: 0.0078


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5939 


######### Eval 286 #########
acc: 0.0078
auc: 0.0078


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6038 


######### Eval 308 #########
acc: 0.0062
auc: 0.0062


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5875 


######### Eval 330 #########
acc: 0.0070
auc: 0.0070


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5772 


######### Eval 352 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5744 


######### Eval 374 #########
acc: 0.0054
auc: 0.0054


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5691 


######### Eval 396 #########
acc: 0.0062
auc: 0.0062


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5820 


######### Eval 418 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5652 


######### Eval 440 #########
acc: 0.0047
auc: 0.0047
save best checkpoint at iter 176 to /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BioClinicalBERT/./best
load best checkpoint at last from /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BioClinicalBERT/./best

✅ BioClinicalBERT complete!
   Time: 0.83h
   Best model saved to: /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BioClinicalBERT/best

🚀 TRAINING: PubMedBERT
✅ BERT_TYPE updated to: microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext

📥 Loading model...

🔒 Freezing encoders (keeping projections trainable)...
   Trainable: 786,433 (0.58%)
   Frozen: 135,829,626 (99.42%)
   Vision projection: 393,216 trainable ✅
   Text projection: 393,216 trainable ✅
   Logit scale: trainable ✅

📊 Creating train/validation split (10% val)...
   Train: 11,569 samples
   Val: 1,285 samples

⚙️ Preparing training...
   Steps per epoch: 22

🎯 Starting training (20 epochs)...
   Batc

Epoch:   0%|          | 0/20 [00:00<?, ?it/s]

Iteration:   0%|          | 0/23 [00:00<?, ?it/s]

/usr/local/lib/python3.12/dist-packages/transformers/models/clip/processing_clip.py:66: FutureWarning: `feature_extractor` is deprecated and will be removed in v5. Use `image_processor` instead.
  warnings.warn(



######### Train Loss #########
0 6.2214 


######### Eval 22 #########
acc: 0.0023
auc: 0.0023


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 6.0933 


######### Eval 44 #########
acc: 0.0008
auc: 0.0008


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.9878 


######### Eval 66 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.9029 


######### Eval 88 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.8188 


######### Eval 110 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7763 


######### Eval 132 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7159 


######### Eval 154 #########
acc: 0.0023
auc: 0.0023


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6978 


######### Eval 176 #########
acc: 0.0016
auc: 0.0016


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6666 


######### Eval 198 #########
acc: 0.0008
auc: 0.0008


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6389 


######### Eval 220 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6338 


######### Eval 242 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6170 


######### Eval 264 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6051 


######### Eval 286 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5920 


######### Eval 308 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5753 


######### Eval 330 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5937 


######### Eval 352 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5859 


######### Eval 374 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5649 


######### Eval 396 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5758 


######### Eval 418 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5673 


######### Eval 440 #########
acc: 0.0039
auc: 0.0039
save best checkpoint at iter 88 to /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/PubMedBERT/./best
load best checkpoint at last from /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/PubMedBERT/./best

✅ PubMedBERT complete!
   Time: 0.83h
   Best model saved to: /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/PubMedBERT/best

🚀 TRAINING: BioBERT
✅ BERT_TYPE updated to: dmis-lab/biobert-base-cased-v1.1

📥 Loading model...

🔒 Freezing encoders (keeping projections trainable)...
   Trainable: 786,433 (0.58%)
   Frozen: 135,829,626 (99.42%)
   Vision projection: 393,216 trainable ✅
   Text projection: 393,216 trainable ✅
   Logit scale: trainable ✅

📊 Creating train/validation split (10% val)...
   Train: 11,569 samples
   Val: 1,285 samples

⚙️ Preparing training...
   Steps per epoch: 22

🎯 Starting training (20 epochs)...
   Batch size: 512
   Learning rate: 6e-05
   Validation eve

Epoch:   0%|          | 0/20 [00:00<?, ?it/s]

Iteration:   0%|          | 0/23 [00:00<?, ?it/s]

/usr/local/lib/python3.12/dist-packages/transformers/models/clip/processing_clip.py:66: FutureWarning: `feature_extractor` is deprecated and will be removed in v5. Use `image_processor` instead.
  warnings.warn(



######### Train Loss #########
0 6.2109 


######### Eval 22 #########
acc: 0.0023
auc: 0.0023


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 6.0725 


######### Eval 44 #########
acc: 0.0016
auc: 0.0016


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.9758 


######### Eval 66 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.8830 


######### Eval 88 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.8054 


######### Eval 110 #########
acc: 0.0023
auc: 0.0023


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7513 


######### Eval 132 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7022 


######### Eval 154 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6855 


######### Eval 176 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6603 


######### Eval 198 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6528 


######### Eval 220 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6113 


######### Eval 242 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6109 


######### Eval 264 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6026 


######### Eval 286 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5811 


######### Eval 308 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5836 


######### Eval 330 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5742 


######### Eval 352 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5677 


######### Eval 374 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5676 


######### Eval 396 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5654 


######### Eval 418 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5687 


######### Eval 440 #########
acc: 0.0047
auc: 0.0047
save best checkpoint at iter 264 to /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BioBERT/./best
load best checkpoint at last from /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BioBERT/./best

✅ BioBERT complete!
   Time: 0.83h
   Best model saved to: /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BioBERT/best

🚀 TRAINING: BERT-base
✅ BERT_TYPE updated to: bert-base-uncased

📥 Loading model...

🔒 Freezing encoders (keeping projections trainable)...
   Trainable: 786,433 (0.58%)
   Frozen: 135,829,626 (99.42%)
   Vision projection: 393,216 trainable ✅
   Text projection: 393,216 trainable ✅
   Logit scale: trainable ✅

📊 Creating train/validation split (10% val)...
   Train: 11,569 samples
   Val: 1,285 samples

⚙️ Preparing training...
   Steps per epoch: 22

🎯 Starting training (20 epochs)...
   Batch size: 512
   Learning rate: 6e-05
   Validation every 22 steps (every epoch

Epoch:   0%|          | 0/20 [00:00<?, ?it/s]

Iteration:   0%|          | 0/23 [00:00<?, ?it/s]

/usr/local/lib/python3.12/dist-packages/transformers/models/clip/processing_clip.py:66: FutureWarning: `feature_extractor` is deprecated and will be removed in v5. Use `image_processor` instead.
  warnings.warn(



######### Train Loss #########
0 6.2168 


######### Eval 22 #########
acc: 0.0023
auc: 0.0023


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 6.0880 


######### Eval 44 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.9867 


######### Eval 66 #########
acc: 0.0047
auc: 0.0047


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.8970 


######### Eval 88 #########
acc: 0.0016
auc: 0.0016


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.8207 


######### Eval 110 #########
acc: 0.0023
auc: 0.0023


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7659 


######### Eval 132 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.7138 


######### Eval 154 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6904 


######### Eval 176 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6606 


######### Eval 198 #########
acc: 0.0054
auc: 0.0054


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6559 


######### Eval 220 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6259 


######### Eval 242 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6133 


######### Eval 264 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.6080 


######### Eval 286 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5976 


######### Eval 308 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5960 


######### Eval 330 #########
acc: 0.0023
auc: 0.0023


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5854 


######### Eval 352 #########
acc: 0.0039
auc: 0.0039


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5699 


######### Eval 374 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5866 


######### Eval 396 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5643 


######### Eval 418 #########
acc: 0.0031
auc: 0.0031


Iteration:   0%|          | 0/23 [00:00<?, ?it/s]


######### Train Loss #########
0 5.5753 


######### Eval 440 #########
acc: 0.0031
auc: 0.0031
save best checkpoint at iter 198 to /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BERT-base/./best
load best checkpoint at last from /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BERT-base/./best

✅ BERT-base complete!
   Time: 0.83h
   Best model saved to: /content/drive/MyDrive/MEDCLIP-PEACEIN/checkpoints/BERT-base/best


🎉 ALL ENCODERS TRAINED!

📊 Summary:
   BioClinicalBERT     : 0.83h
   PubMedBERT          : 0.83h
   BioBERT             : 0.83h
   BERT-base           : 0.83h

   Total: 3.31h


---

# Section 7: Zero-Shot Evaluation 📊

Evaluate all trained models on the test set.

In [1]:
import sys
sys.path.insert(0, MEDCLIP_DIR)

from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, classification_report
from medclip import PromptClassifier
import numpy as np

evaluation_results = {}

for encoder_config in TEXT_ENCODERS:
    encoder_name = encoder_config['name']

    print(f"\n{'='*80}")
    print(f"📊 EVALUATING: {encoder_name}")
    print(f"{'='*80}")

    # Load trained model
    model_path = training_results[encoder_name]['checkpoint_path']
    print(f"\n📥 Loading model from {model_path}")
    model = MedCLIPModel.from_pretrained(model_path)
    model.cuda()
    model.eval()

    # Create prompt classifier
    classifier = PromptClassifier(
        model=model,
        labels=DISEASE_LABELS,
        templates=[
            "Fundus photograph showing {}.",
            "Color fundus image demonstrating {}.",
            "Retinal imaging reveals {}."
        ]
    )

    # Load ground truth labels from test_texts.jsonl
    print(f"\n📋 Loading ground truth labels...")
    ground_truth_labels = []
    with open(test_texts_jsonl) as f:
        for line in f:
            entry = json.loads(line)
            ground_truth_labels.append(entry['label'])

    print(f"   Ground truth samples: {len(ground_truth_labels)}")

    # Evaluate
    print(f"\n🧪 Running evaluation on {len(test_dataset):,} samples...")
    test_dataloader = DataLoader(test_dataset, batch_size=64,
                                  shuffle=False, num_workers=2)

    predictions = []

    with torch.no_grad():
        for i, batch in enumerate(test_dataloader):
            if i % 10 == 0:
                print(f"   Batch {i}/{len(test_dataloader)}")

            images = batch['pixel_values'].cuda()
            preds = classifier(images)
            predictions.extend(preds)

    # Convert predictions and ground truth to indices
    label_to_idx = {label: idx for idx, label in enumerate(DISEASE_LABELS)}

    # Map predictions to indices
    pred_indices = [label_to_idx.get(pred, 0) for pred in predictions]

    # Ensure same length
    if len(pred_indices) != len(ground_truth_labels):
        print(f"\n⚠️ Warning: Predictions ({len(pred_indices)}) != Ground truth ({len(ground_truth_labels)})")
        min_len = min(len(pred_indices), len(ground_truth_labels))
        pred_indices = pred_indices[:min_len]
        ground_truth_labels = ground_truth_labels[:min_len]

    # Compute metrics
    accuracy = accuracy_score(ground_truth_labels, pred_indices)
    f1_macro = f1_score(ground_truth_labels, pred_indices, average='macro', zero_division=0)
    f1_weighted = f1_score(ground_truth_labels, pred_indices, average='weighted', zero_division=0)

    print(f"\n✅ Evaluation complete")
    print(f"   Accuracy: {accuracy*100:.2f}%")
    print(f"   F1 Macro: {f1_macro:.4f}")
    print(f"   F1 Weighted: {f1_weighted:.4f}")

    evaluation_results[encoder_name] = {
        'predictions': pred_indices,
        'ground_truth': ground_truth_labels,
        'accuracy': accuracy,
        'f1_macro': f1_macro,
        'f1_weighted': f1_weighted
    }

    del model, classifier
    torch.cuda.empty_cache()

print(f"\n🎉 ALL EVALUATIONS COMPLETE!")

NameError: name 'MEDCLIP_DIR' is not defined

---

# Section 8: Results Comparison 📊

Compare performance across all encoders.

## 8.1 Summary Table

In [ ]:
import pandas as pd
import numpy as np

print("="*80)
print("📊 RESULTS SUMMARY")
print("="*80)

# Compile results
results_data = []
for encoder_name, train_result in training_results.items():
    eval_result = evaluation_results.get(encoder_name, {})

    results_data.append({
        'Encoder': encoder_name,
        'Training Time (h)': f"{train_result['training_time_hours']:.2f}",
        'Accuracy (%)': f"{eval_result.get('accuracy', 0)*100:.2f}" if 'accuracy' in eval_result else 'N/A',
        'F1 Macro': f"{eval_result.get('f1_macro', 0):.4f}" if 'f1_macro' in eval_result else 'N/A',
        'F1 Weighted': f"{eval_result.get('f1_weighted', 0):.4f}" if 'f1_weighted' in eval_result else 'N/A',
    })

results_df = pd.DataFrame(results_data)

print("\n" + results_df.to_string(index=False))
print("\n" + "="*80)

# Save to CSV
results_csv = f"{DRIVE_RESULTS}/encoder_comparison.csv"
results_df.to_csv(results_csv, index=False)
print(f"\n✅ Results saved to {results_csv}")

## 8.2 Visualization: Accuracy Comparison

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style("whitegrid")

# Extract accuracy values
encoder_names = [r['Encoder'] for r in results_data]
accuracies = [eval_result.get('accuracy', 0)*100 for encoder_name, eval_result in
              [(e, evaluation_results.get(e, {})) for e in encoder_names]]

# Create bar plot
fig, ax = plt.subplots(figsize=(10, 6))
bars = ax.bar(encoder_names, accuracies, color=['#2ecc71', '#3498db', '#9b59b6', '#e74c3c'])

# Customize
ax.set_ylabel('Accuracy (%)', fontsize=12)
ax.set_xlabel('Text Encoder', fontsize=12)
ax.set_title('MedCLIP Performance Comparison: Frozen Encoders on Peacein Dataset',
             fontsize=14, fontweight='bold')
ax.set_ylim([0, 100])

# Add value labels on bars
for bar in bars:
    height = bar.get_height()
    ax.text(bar.get_x() + bar.get_width()/2., height,
            f'{height:.1f}%',
            ha='center', va='bottom', fontsize=10, fontweight='bold')

# Rotate x labels
plt.xticks(rotation=15, ha='right')
plt.tight_layout()

# Save
plot_path = f"{DRIVE_RESULTS}/accuracy_comparison.png"
plt.savefig(plot_path, dpi=300, bbox_inches='tight')
print(f"✅ Plot saved to {plot_path}")
plt.show()

## 8.3 F1 Score Comparison

In [ ]:
# Extract F1 scores
f1_macro = [evaluation_results.get(e, {}).get('f1_macro', 0) for e in encoder_names]
f1_weighted = [evaluation_results.get(e, {}).get('f1_weighted', 0) for e in encoder_names]

# Create grouped bar plot
fig, ax = plt.subplots(figsize=(10, 6))

x = np.arange(len(encoder_names))
width = 0.35

bars1 = ax.bar(x - width/2, f1_macro, width, label='F1 Macro', color='#3498db', alpha=0.8)
bars2 = ax.bar(x + width/2, f1_weighted, width, label='F1 Weighted', color='#2ecc71', alpha=0.8)

ax.set_ylabel('F1 Score', fontsize=12)
ax.set_xlabel('Text Encoder', fontsize=12)
ax.set_title('F1 Score Comparison: Macro vs Weighted', fontsize=14, fontweight='bold')
ax.set_xticks(x)
ax.set_xticklabels(encoder_names, rotation=15, ha='right')
ax.legend()
ax.set_ylim([0, 1.0])

# Add value labels
for bars in [bars1, bars2]:
    for bar in bars:
        height = bar.get_height()
        ax.text(bar.get_x() + bar.get_width()/2., height,
                f'{height:.3f}',
                ha='center', va='bottom', fontsize=9)

plt.tight_layout()

# Save
plot_path = f"{DRIVE_RESULTS}/f1_comparison.png"
plt.savefig(plot_path, dpi=300, bbox_inches='tight')
print(f"✅ Plot saved to {plot_path}")
plt.show()

---

# Section 9: Per-Class Analysis 🔬

Analyze performance on each disease class.

## 9.1 Confusion Matrices

In [ ]:
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns

print("="*80)
print("🔬 PER-CLASS ANALYSIS")
print("="*80)

fig, axes = plt.subplots(2, 2, figsize=(16, 14))
axes = axes.flatten()

for idx, (encoder_name, eval_result) in enumerate(
    [(e, evaluation_results.get(e, {})) for e in encoder_names]
):
    if 'predictions' not in eval_result or 'ground_truth' not in eval_result:
        print(f"\n⚠️ Skipping {encoder_name}: Missing predictions/ground truth")
        continue

    # Get confusion matrix
    y_true = eval_result['ground_truth']
    y_pred = eval_result['predictions']
    cm = confusion_matrix(y_true, y_pred)

    # Normalize by row (true labels)
    cm_norm = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]

    # Plot
    ax = axes[idx]
    sns.heatmap(cm_norm, annot=False, fmt='.2f', cmap='Blues',
                xticklabels=DISEASE_LABELS, yticklabels=DISEASE_LABELS,
                cbar_kws={'label': 'Normalized Count'}, ax=ax)

    ax.set_title(f'{encoder_name}', fontsize=12, fontweight='bold')
    ax.set_xlabel('Predicted', fontsize=10)
    ax.set_ylabel('True', fontsize=10)
    ax.tick_params(axis='both', labelsize=7)

plt.tight_layout()

# Save
plot_path = f"{DRIVE_RESULTS}/confusion_matrices.png"
plt.savefig(plot_path, dpi=300, bbox_inches='tight')
print(f"\n✅ Confusion matrices saved to {plot_path}")
plt.show()

## 9.2 Per-Class F1 Scores

In [ ]:
from sklearn.metrics import classification_report

# Create per-class comparison
per_class_results = {}

for encoder_name in encoder_names:
    eval_result = evaluation_results.get(encoder_name, {})

    if 'predictions' not in eval_result or 'ground_truth' not in eval_result:
        continue

    y_true = eval_result['ground_truth']
    y_pred = eval_result['predictions']

    # Get classification report
    report = classification_report(y_true, y_pred, target_names=DISEASE_LABELS,
                                    output_dict=True, zero_division=0)

    per_class_results[encoder_name] = report

# Create comparison DataFrame
if per_class_results:
    # Extract F1 scores per class
    per_class_data = []
    for disease in DISEASE_LABELS:
        row = {'Disease': disease}
        for encoder_name in encoder_names:
            if encoder_name in per_class_results:
                f1 = per_class_results[encoder_name].get(disease, {}).get('f1-score', 0)
                row[encoder_name] = f"{f1:.3f}"
        per_class_data.append(row)

    per_class_df = pd.DataFrame(per_class_data)

    print("\n" + "="*80)
    print("📋 PER-CLASS F1 SCORES")
    print("="*80)
    print("\n" + per_class_df.to_string(index=False))

    # Save
    per_class_csv = f"{DRIVE_RESULTS}/per_class_f1.csv"
    per_class_df.to_csv(per_class_csv, index=False)
    print(f"\n✅ Saved to {per_class_csv}")
else:
    print("\n⚠️ No per-class results available")

---

# Section 10: Statistical Analysis 📈

Test for significant differences between encoders.

## 10.1 Training Time Analysis

In [ ]:
print("="*80)
print("⏱️ TRAINING TIME ANALYSIS")
print("="*80)

# Training time comparison
train_times = [training_results[e]['training_time_hours'] for e in encoder_names]

print(f"\nMean training time: {np.mean(train_times):.2f}h")
print(f"Std training time: {np.std(train_times):.2f}h")
print(f"Min: {np.min(train_times):.2f}h ({encoder_names[np.argmin(train_times)]})")
print(f"Max: {np.max(train_times):.2f}h ({encoder_names[np.argmax(train_times)]})")

# Create visualization
fig, ax = plt.subplots(figsize=(10, 6))
bars = ax.bar(encoder_names, train_times, color='#e67e22', alpha=0.7)

ax.set_ylabel('Training Time (hours)', fontsize=12)
ax.set_xlabel('Text Encoder', fontsize=12)
ax.set_title('Training Time Comparison (Frozen Encoders)', fontsize=14, fontweight='bold')

# Add value labels
for bar in bars:
    height = bar.get_height()
    ax.text(bar.get_x() + bar.get_width()/2., height,
            f'{height:.2f}h',
            ha='center', va='bottom', fontsize=10)

plt.xticks(rotation=15, ha='right')
plt.tight_layout()

# Save
plot_path = f"{DRIVE_RESULTS}/training_time_comparison.png"
plt.savefig(plot_path, dpi=300, bbox_inches='tight')
print(f"\n✅ Plot saved to {plot_path}")
plt.show()

print("\n" + "="*80)

## 10.2 Key Findings Summary

In [ ]:
print("="*80)
print("🎯 KEY FINDINGS")
print("="*80)

# Find best performer
best_idx = np.argmax(accuracies)
best_encoder = encoder_names[best_idx]
best_accuracy = accuracies[best_idx]

# Find baseline (BERT-base)
baseline_idx = encoder_names.index('BERT-base') if 'BERT-base' in encoder_names else None

findings = []

findings.append(f"\n1. BEST PERFORMER:")
findings.append(f"   → {best_encoder}: {best_accuracy:.2f}% accuracy")

if baseline_idx is not None:
    baseline_accuracy = accuracies[baseline_idx]
    improvement = best_accuracy - baseline_accuracy
    findings.append(f"\n2. MEDICAL PRETRAINING BENEFIT:")
    findings.append(f"   → {improvement:+.2f}% over general BERT baseline")
    findings.append(f"   → Medical encoders {'DO' if improvement > 0 else 'DO NOT'} improve performance")

findings.append(f"\n3. TRAINING EFFICIENCY:")
findings.append(f"   → Mean training time: {np.mean(train_times):.2f}h per encoder")
findings.append(f"   → Total time (4 encoders): {np.sum(train_times):.2f}h")
findings.append(f"   → Frozen encoder approach is {'efficient' if np.mean(train_times) < 3 else 'moderate'}")

if per_class_results:
    findings.append(f"\n4. DISEASE DETECTION:")
    # Find hardest disease (lowest average F1)
    disease_avg_f1 = {}
    for disease in DISEASE_LABELS:
        f1_scores = [per_class_results[e].get(disease, {}).get('f1-score', 0)
                     for e in per_class_results.keys()]
        disease_avg_f1[disease] = np.mean(f1_scores)

    easiest = max(disease_avg_f1, key=disease_avg_f1.get)
    hardest = min(disease_avg_f1, key=disease_avg_f1.get)

    findings.append(f"   → Easiest to detect: {easiest} (F1: {disease_avg_f1[easiest]:.3f})")
    findings.append(f"   → Hardest to detect: {hardest} (F1: {disease_avg_f1[hardest]:.3f})")

findings.append(f"\n5. RECOMMENDATION:")
if best_encoder == 'BioClinicalBERT':
    findings.append(f"   → Use MedCLIP's default BioClinicalBERT encoder")
else:
    findings.append(f"   → Consider {best_encoder} over default BioClinicalBERT")

for finding in findings:
    print(finding)

# Save findings
findings_file = f"{DRIVE_RESULTS}/key_findings.txt"
with open(findings_file, 'w') as f:
    f.write('\n'.join(findings))

print(f"\n✅ Findings saved to {findings_file}")
print("\n" + "="*80)

## 10.3 Export Complete Results

In [ ]:
# Final export
final_results = {
    'training': training_results,
    'evaluation': evaluation_results,
    'summary': {
        'encoder_names': encoder_names,
        'accuracies': accuracies,
        'f1_macro': f1_macro,
        'f1_weighted': f1_weighted,
        'train_times': train_times,
        'best_encoder': best_encoder,
        'best_accuracy': float(best_accuracy)
    },
    'per_class': per_class_results if per_class_results else {}
}

results_json = f"{DRIVE_RESULTS}/complete_results.json"
with open(results_json, 'w') as f:
    json.dump(final_results, f, indent=2, default=str)

print("✅ Complete results saved!")
print(f"   Location: {results_json}")

print("\n" + "="*80)
print("🎉 ANALYSIS COMPLETE!")
print("="*80)
print("\nGenerated files:")
print(f"   1. {DRIVE_RESULTS}/encoder_comparison.csv")
print(f"   2. {DRIVE_RESULTS}/accuracy_comparison.png")
print(f"   3. {DRIVE_RESULTS}/f1_comparison.png")
print(f"   4. {DRIVE_RESULTS}/confusion_matrices.png")
print(f"   5. {DRIVE_RESULTS}/per_class_f1.csv")
print(f"   6. {DRIVE_RESULTS}/training_time_comparison.png")
print(f"   7. {DRIVE_RESULTS}/key_findings.txt")
print(f"   8. {DRIVE_RESULTS}/complete_results.json")
print("\n" + "="*80)

---

# 📚 References

## Primary Papers

1. **MedCLIP: Contrastive Learning from Unpaired Medical Images and Text**  
   Wang, Z., Wu, Z., Agarwal, D., & Sun, J. (2022).  
   *arXiv preprint arXiv:2210.10163*  
   [https://arxiv.org/abs/2210.10163](https://arxiv.org/abs/2210.10163)  
   - Primary method used in this notebook
   - Contrastive learning framework for medical vision-language alignment

2. **Learning Transferable Visual Models From Natural Language Supervision (CLIP)**  
   Radford, A., Kim, J. W., Hallacy, C., et al. (2021).  
   *International Conference on Machine Learning (ICML)*  
   [https://arxiv.org/abs/2103.00020](https://arxiv.org/abs/2103.00020)  
   - Foundation for contrastive vision-language learning
   - Inspired MedCLIP architecture

## Vision Encoder

3. **Swin Transformer: Hierarchical Vision Transformer using Shifted Windows**  
   Liu, Z., Lin, Y., Cao, Y., et al. (2021).  
   *IEEE/CVF International Conference on Computer Vision (ICCV)*  
   [https://arxiv.org/abs/2103.14030](https://arxiv.org/abs/2103.14030)  
   - Vision backbone used in MedCLIP: `microsoft/swin-tiny-patch4-window7-224`
   - Hierarchical transformer with shifted window attention

## Text Encoders

4. **BioClinicalBERT: Publicly Available Clinical BERT Embeddings**  
   Alsentzer, E., Murphy, J., Boag, W., et al. (2019).  
   *Proceedings of the 2nd Clinical Natural Language Processing Workshop*  
   [https://arxiv.org/abs/1904.03323](https://arxiv.org/abs/1904.03323)  
   - Default MedCLIP encoder: `emilyalsentzer/Bio_ClinicalBERT`
   - Pretrained on MIMIC-III clinical notes (2 million notes)

5. **Domain-Specific Language Model Pretraining for Biomedical Natural Language Processing (PubMedBERT)**  
   Gu, Y., Tinn, R., Cheng, H., et al. (2021).  
   *ACM Transactions on Computing for Healthcare (HEALTH)*  
   [https://arxiv.org/abs/2007.15779](https://arxiv.org/abs/2007.15779)  
   - Model ID: `microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext`
   - Pretrained from scratch on PubMed abstracts (14M) and full-text articles (3M)

6. **BioBERT: a pre-trained biomedical language representation model**  
   Lee, J., Yoon, W., Kim, S., et al. (2019).  
   *Bioinformatics, 36(4), 1234-1240*  
   [https://arxiv.org/abs/1901.08746](https://arxiv.org/abs/1901.08746)  
   - Model ID: `dmis-lab/biobert-base-cased-v1.1`
   - BERT fine-tuned on PubMed abstracts (4.5B words) + PMC full-text (13.5B words)

7. **BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding**  
   Devlin, J., Chang, M. W., Lee, K., & Toutanova, K. (2019).  
   *Proceedings of NAACL-HLT*  
   [https://arxiv.org/abs/1810.04805](https://arxiv.org/abs/1810.04805)  
   - Baseline encoder: `bert-base-uncased`
   - Pretrained on Wikipedia (2.5B words) and BookCorpus (800M words)

## Dataset

8. **Peacein Color Fundus Eye Dataset**  
   Peacein Innovations (2024).  
   Available on HuggingFace: `Peacein/color-fundus-eye`  
   [https://huggingface.co/datasets/Peacein/color-fundus-eye](https://huggingface.co/datasets/Peacein/color-fundus-eye)  
   
   **Dataset Statistics**:
   - **Total Images**: 16,242 color fundus photographs
   - **Train Split**: 12,989 samples (80%)
   - **Test Split**: 3,253 samples (20%)
   - **Image Format**: RGB color fundus photography
   - **Resolution**: Variable (resized to 224×224 for training)
   - **Classes**: 10 retinal conditions
   
   **Disease Distribution**:
   1. Central Serous Chorioretinopathy (CSCR)
   2. Diabetic Retinopathy (DR)
   3. Disc Edema
   4. Glaucoma
   5. Healthy (normal fundus)
   6. Macular Scar
   7. Myopia (high myopia with pathological changes)
   8. Pterygium
   9. Retinal Detachment
   10. Retinitis Pigmentosa
   
   **Data Collection**:
   - Curated from clinical ophthalmology databases
   - Quality-controlled fundus photography
   - Expert-labeled diagnoses
   - De-identified patient information
   
   **Related Datasets**:
   - Similar to ODIR-5K, but with different disease distribution
   - Complementary to RFMiD (Retinal Fundus Multi-disease Image Dataset)
   
   **Usage License**: Apache 2.0 (check HuggingFace dataset card for updates)
   
   **Citation** (if using this dataset):
   ```
   @misc{peacein2024fundus,
     title={Peacein Color Fundus Eye Dataset},
     author={Peacein Innovations},
     year={2024},
     publisher={HuggingFace},
     howpublished={\url{https://huggingface.co/datasets/Peacein/color-fundus-eye}}
   }
   ```

## Training Methodology

9. **Linear Probing / Feature Extraction**  
   Kumar, A., Raghunathan, A., Jones, R., et al. (2022).  
   *Fine-Tuning can Distort Pretrained Features and Underperform Out-of-Distribution*  
   *ICLR 2022*  
   [https://arxiv.org/abs/2202.10054](https://arxiv.org/abs/2202.10054)  
   - Justification for frozen encoder training
   - Preserves pretrained representations
   - Often outperforms full fine-tuning on small datasets

## Loss Function

10. **Contrastive Loss (InfoNCE)**  
    van den Oord, A., Li, Y., & Vinyals, O. (2018).  
    *Representation Learning with Contrastive Predictive Coding*  
    *arXiv preprint arXiv:1807.03748*  
    [https://arxiv.org/abs/1807.03748](https://arxiv.org/abs/1807.03748)  
    - Used in MedCLIP's `ImageTextContrastiveLoss`
    - Aligns image-text pairs in shared embedding space
    - Normalized temperature-scaled cross entropy (NT-Xent)

---

## BibTeX Citations

```bibtex
@article{wang2022medclip,
  title={MedCLIP: Contrastive Learning from Unpaired Medical Images and Text},
  author={Wang, Zifeng and Wu, Zhenbang and Agarwal, Dinesh and Sun, Jimeng},
  journal={arXiv preprint arXiv:2210.10163},
  year={2022}
}

@inproceedings{radford2021clip,
  title={Learning Transferable Visual Models From Natural Language Supervision},
  author={Radford, Alec and Kim, Jong Wook and Hallacy, Chris and others},
  booktitle={International Conference on Machine Learning},
  year={2021}
}

@inproceedings{liu2021swin,
  title={Swin Transformer: Hierarchical Vision Transformer using Shifted Windows},
  author={Liu, Ze and Lin, Yutong and Cao, Yue and others},
  booktitle={IEEE/CVF International Conference on Computer Vision},
  pages={10012--10022},
  year={2021}
}

@inproceedings{alsentzer2019bioclinicalbert,
  title={Publicly Available Clinical BERT Embeddings},
  author={Alsentzer, Emily and Murphy, John and Boag, William and others},
  booktitle={Proceedings of the 2nd Clinical Natural Language Processing Workshop},
  pages={72--78},
  year={2019}
}

@article{gu2021pubmedbert,
  title={Domain-Specific Language Model Pretraining for Biomedical Natural Language Processing},
  author={Gu, Yu and Tinn, Robert and Cheng, Hao and others},
  journal={ACM Transactions on Computing for Healthcare},
  volume={3},
  number={1},
  pages={1--23},
  year={2021}
}

@article{lee2019biobert,
  title={BioBERT: a pre-trained biomedical language representation model},
  author={Lee, Jinhyuk and Yoon, Wonjin and Kim, Sungdong and others},
  journal={Bioinformatics},
  volume={36},
  number={4},
  pages={1234--1240},
  year={2020}
}

@inproceedings{devlin2019bert,
  title={BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding},
  author={Devlin, Jacob and Chang, Ming-Wei and Lee, Kenton and Toutanova, Kristina},
  booktitle={Proceedings of NAACL-HLT},
  pages={4171--4186},
  year={2019}
}

@misc{peacein2024fundus,
  title={Peacein Color Fundus Eye Dataset},
  author={Peacein Innovations},
  year={2024},
  publisher={HuggingFace},
  howpublished={\url{https://huggingface.co/datasets/Peacein/color-fundus-eye}}
}

@inproceedings{kumar2022finetuning,
  title={Fine-Tuning can Distort Pretrained Features and Underperform Out-of-Distribution},
  author={Kumar, Ananya and Raghunathan, Aditi and Jones, Robbie and others},
  booktitle={International Conference on Learning Representations},
  year={2022}
}

@article{oord2018cpc,
  title={Representation Learning with Contrastive Predictive Coding},
  author={van den Oord, Aaron and Li, Yazhe and Vinyals, Oriol},
  journal={arXiv preprint arXiv:1807.03748},
  year={2018}
}
```

---

## Software & Libraries

- **PyTorch**: Paszke, A., et al. (2019). PyTorch: An Imperative Style, High-Performance Deep Learning Library. *NeurIPS*.  
- **HuggingFace Transformers**: Wolf, T., et al. (2020). Transformers: State-of-the-Art Natural Language Processing. *EMNLP: System Demonstrations*.  
- **HuggingFace Datasets**: Lhoest, Q., et al. (2021). Datasets: A Community Library for Natural Language Processing. *EMNLP*.  
- **DSPy**: Khattab, O., et al. (2023). DSPy: Compiling Declarative Language Model Calls into Self-Improving Pipelines. *arXiv:2310.03714*.  
- **Scikit-learn**: Pedregosa, F., et al. (2011). Scikit-learn: Machine Learning in Python. *JMLR*.

---

## Related Work on Retinal Disease Classification

- **ODIR-5K**: Retinal fundus multi-label classification dataset (Peking University)
- **RFMiD**: Retinal Fundus Multi-disease Image Dataset (IEEE DataPort)
- **Messidor-2**: Diabetic retinopathy grading dataset
- **EyePACS**: Large-scale diabetic retinopathy dataset (Kaggle)
- **APTOS 2019**: Asia Pacific Tele-Ophthalmology Society blindness detection

---

**Last Updated**: 2025-11-24  
**Notebook Version**: 1.0  
**Maintained by**: [Your Name/Lab]  
